### PR TITLE
Akka Persistence instrumentation improvements

### PR DESCRIPTION
**Pull request by ygree**
## Purpose

Internal improvements in `Eventsourced` to simplify instrumentation.

### CODE REVIEW

**Comment by ygree** [comment]
@patriknw @chbatey @raboof I've made a couple of improvements to simplify instrumentation. I used MiMa plugin to verify binary compatibility. Would be great if we could add it to the next 2.5.x release.

**Comment by johanandren** [comment]
OK TO TEST

**Comment by johanandren** [comment]
PLS WHITELIST

**Comment by chbatey** [comment]
I don't think MiMa is sufficient so let's not merge this yet but I think @raboof  will look into this next week

**Comment by ygree** [comment]
> I don't think MiMa is sufficient so let's not merge this yet but I think @raboof will look into this next week

I took a look at the resulting classes and the differences:

In Scala 2.11 Eventsourced interface stays the same and it only adds next extra methods to Eventsourced$class:

```java
  private static void sendBatchedEventsToJournal(akka.persistence.Eventsourced, scala.collection.immutable.Vector);
  private static void batchAtomicWrite(akka.persistence.Eventsourced, akka.persistence.AtomicWrite);
  public static void akka$persistence$Eventsourced$$writeEventSucceeded(akka.persistence.Eventsourced, akka.persistence.PersistentRepr);
  public static void akka$persistence$Eventsourced$$writeEventRejected(akka.persistence.Eventsourced, akka.persistence.PersistentRepr, java.lang.Throwable);
  public static void akka$persistence$Eventsourced$$writeEventFailed(akka.persistence.Eventsourced, akka.persistence.PersistentRepr, java.lang.Throwable);
```

In Scala 2.12 it adds next extra methods to Eventsourced:

```java
  private void sendBatchedEventsToJournal(scala.collection.immutable.Vector<akka.persistence.PersistentEnvelope>);
  private void batchAtomicWrite(akka.persistence.AtomicWrite);
  public static void akka$persistence$Eventsourced$$writeEventSucceeded$(akka.persistence.Eventsourced, akka.persistence.PersistentRepr);
  public void akka$persistence$Eventsourced$$writeEventSucceeded(akka.persistence.PersistentRepr);
  public static void akka$persistence$Eventsourced$$writeEventRejected$(akka.persistence.Eventsourced, akka.persistence.PersistentRepr, java.lang.Throwable);
  public void akka$persistence$Eventsourced$$writeEventRejected(akka.persistence.PersistentRepr, java.lang.Throwable);
  public static void akka$persistence$Eventsourced$$writeEventFailed$(akka.persistence.Eventsourced, akka.persistence.PersistentRepr, java.lang.Throwable);
  public void akka$persistence$Eventsourced$$writeEventFailed(akka.persistence.PersistentRepr, java.lang.Throwable);
```

In Scala 2.13 it adds next extra methods to Eventsourced:

```java
  private void sendBatchedEventsToJournal(scala.collection.immutable.Vector<akka.persistence.PersistentEnvelope>);
  private void batchAtomicWrite(akka.persistence.AtomicWrite);
  public static void akka$persistence$Eventsourced$$writeEventSucceeded$(akka.persistence.Eventsourced, akka.persistence.PersistentRepr);
  public void akka$persistence$Eventsourced$$writeEventSucceeded(akka.persistence.PersistentRepr);
  public static void akka$persistence$Eventsourced$$writeEventRejected$(akka.persistence.Eventsourced, akka.persistence.PersistentRepr, java.lang.Throwable);
  public void akka$persistence$Eventsourced$$writeEventRejected(akka.persistence.PersistentRepr, java.lang.Throwable);
  public static void akka$persistence$Eventsourced$$writeEventFailed$(akka.persistence.Eventsourced, akka.persistence.PersistentRepr, java.lang.Throwable);
  public void akka$persistence$Eventsourced$$writeEventFailed(akka.persistence.PersistentRepr, java.lang.Throwable);
```

Should not be a problem, since we are only adding new private methods that become private methods or mingled public and public static methods. What do you think? Should we take an extra step and actually verify it with some example? How do you usually verify binary compatibility?

File: akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala [hunk 1]
```diff
@@ -791,6 +803,21 @@ private[persistence] trait Eventsourced
...
+
+  @InternalStableApi
+  private def writeEventRejected(p: PersistentRepr, cause: Throwable): Unit = {
+    onPersistRejected(cause, p.payload, p.sequenceNr)
+  }
+
+  @InternalStableApi
+  private def writeEventFailed(p: PersistentRepr, cause: Throwable): Unit = {
+    onPersistFailure(cause, p.payload, p.sequenceNr)
+  }
```
Comment (helena) [inline akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala]: Unimportant nit but one liners don't really need `{..}`

**Comment by chbatey** [comment]
We should probably try a sample library with this just in case. It's main benefit is being back ported to 2.5.x so no rush for M2

**Review by helena — APPROVED** [review]
LGTM no questions just a nit

**Review by raboof — COMMENTED** [review]
Looks convincing, and the analysis that this should be binary compatible also looks solid, but let's verify it anyway as suggested.

**Comment by patriknw** [comment]
I think it's more important to look with `:javap` of a concrete class extending `PersistentActor`, for example:
```
sbt:akka-persistence> test:console
:javap -p akka.persistence.ManyRecoveriesSpec$TestPersistentActor
```

Such class must not include these new methods, because if Akka 2.5.24 calls those and they don't exist in a class from a library compiled with 2.5.23 it will boom.

I looked at that with Scala 2.11, which is most troublesome.
It's looking good. These private methods are not added there. Compared to fields which are, e.g. `currentState`.

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**